test/grammar/doctypedecl.test.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+const { describe, expect, test } = require('@jest/globals');
+const { doctypedecl_uv, Name, S, ExternalID, QName, SystemLiteral } = require('../../lib/grammar');
+const fs = require('fs');
+var REGEXP_DUMP = `'use strict';
+// THIS FILE IS GENERATED by tests, don't change it manually
+${Object.entries({
+	S,
+	Name,
+	QName,
+	SystemLiteral,
+	ExternalID,
+	doctypedecl_uv,
+})
+	.map(([name, reg]) => `const ${name} = /${reg.source}/;`)
+	.join('\n')}`;
+describe('all relevant regular expressions written out to a file', () => {
+	test('should be updated when running tests', () => {
+		var fileName = __dirname + '/regexp.js';
+		// delete the file and rerun the test(s) to update to current value, in case you touched grammar.js
+		if (!fs.existsSync(fileName)) {
+			fs.writeFileSync(fileName, REGEXP_DUMP);
+		}
+		expect(fs.readFileSync(fileName, 'utf-8')).toBe(REGEXP_DUMP);
+	});
+})
+describe('doctypedecl_uv', () => {
+	test('should match parts', () => {
+		expect(Name.test('name')).toBe(true);
+		expect(SystemLiteral.test('""')).toBe(true);
+		expect(SystemLiteral.test("''")).toBe(true);
+		expect(ExternalID.test(`PUBLIC '' ''`)).toBe(true);
+		expect(ExternalID.test(`SYSTEM ''`)).toBe(true);
+		expect(ExternalID.test(`SYSTEM ""`)).toBe(true);
+		expect(ExternalID.test(`SYSTEM "hello.dtd"`)).toBe(true);
+	});
+	[
+		`<!DOCTYPE name PUBLIC "identifier"  "url" [
+		<?foo "foo"?>
+		<!ENTITY bar "bar">
+		<!ENTITY bar "bar2"> <!--  -->
+		<!ENTITY % baz "baz"> %test;
+		]>`,
+		`<!DOCTYPE name>`,
+		`<!DOCTYPE name SYSTEM "identifier" [
+		  <!ENTITY foo "foo">
+		  <!ENTITY bar "bar">
+		  <!ENTITY bar "bar2"> <!--  -->
+		  <!ENTITY % baz "baz"> %test;
+		]>`,
+		`<!DOCTYPE name [
+		 <!ELEMENT br EMPTY>
+		 <!ELEMENT p (#PCDATA|emph)* >
+		 <!ELEMENT container ANY>
+		 ]>`,
+		/*
+		TODO: This example comes from the documentation,
+		      but elementdecl only allows `Name` after the first whitespace which doesn't allow % or ;.
+		      Does this mean we need to do entity replacement in elements?
+		 <!ELEMENT %name.para; %content.para; >
```

Comment:
Not sure what do to with this example, even though it comes from the spec, the Grammar looks like this shouldn't be allowed.